Task 3


In [ ]:
import pandas as pd
link = "https://storage.googleapis.com/dataset-uploader/bbc/bbc-text.csv"
df = pd.read_csv(link)

Data Info

In [ ]:
print("The total number of documents in the dataset:", len(df))
print("The total number of documents per class:")
print(df['category'].value_counts())
print("One sample document per class")
for c in df["category"].unique():
    print(f"\nClass = {c} ")
    print(df[df["category"] == c]["text"].iloc[0][:200])




The total number of documents in the dataset: 2225
The total number of documents per class:
category
sport            511
business         510
politics         417
tech             401
entertainment    386
Name: count, dtype: int64
One sample document per class

Class = tech 
tv future in the hands of viewers with home theatre systems  plasma high-definition tvs  and digital video recorders moving into the living room  the way people watch tv will be radically different in

Class = business 
worldcom boss  left books alone  former worldcom boss bernie ebbers  who is accused of overseeing an $11bn (£5.8bn) fraud  never made accounting decisions  a witness has told jurors.  david myers made

Class = sport 
tigers wary of farrell  gamble  leicester say they will not be rushed into making a bid for andy farrell should the great britain rugby league captain decide to switch codes.   we and anybody else inv

Class = entertainment 
ocean s twelve raids box office ocean s twelve  the crime cap

The document is not that well balanced from the observation above, as sports has the highest no of documents,511, while Entertainment has only 386 documents

Train,Text split

In [ ]:
def  train_test_split(data,t_ratio=0.2):
    size = int(len(data) * t_ratio)
    test = data[:size]
    train = data[size:]
    print(test.head(5))
    print(train.head(5))
    return train, test

train_data,test_data=train_test_split(df,0.2)
training=list(zip(train_data['text'], train_data['category']))
testing=list(zip(test_data['text'], test_data['category']))
print(len(training))
print(len(testing))

        category                                               text
0           tech  tv future in the hands of viewers with home th...
1       business  worldcom boss  left books alone  former worldc...
2          sport  tigers wary of farrell  gamble  leicester say ...
3          sport  yeading face newcastle in fa cup premiership s...
4  entertainment  ocean s twelve raids box office ocean s twelve...
     category                                               text
445  politics  hunt ban support is  in decline  support for a...
446      tech  broadband challenges tv viewing the number of ...
447  business  weak data buffets french economy a batch of do...
448     sport  collins calls for chambers return world 100m c...
449  business  nasdaq planning $100m-share sale the owner of ...
1780
445


Building the model

In [ ]:
import nltk
nltk.download('punkt_tab')
from textblob.classifiers import NaiveBayesClassifier
classifier = NaiveBayesClassifier(training)

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


Accuracy

In [ ]:
classifier.accuracy(testing)

0.9595505617977528

2-fold cross validation

In [ ]:
from sklearn.model_selection import KFold
def cross_validation_split(dataset, k):
    kf = KFold(n_splits=k, shuffle=False, random_state=None)
    kfolds = []
    for train_idx, test_idx in kf.split(dataset):
        train_df = dataset.iloc[train_idx]
        test_df = dataset.iloc[test_idx]
        kfolds.append((train_df, test_df))
    return kfolds


kfolddata = cross_validation_split(df, k=2)
for i, (train_df, test_df) in enumerate(kfolddata):
    print(f"\nFold {i+1}")
    print("Train size:", len(train_df))
    print("Test size:", len(test_df))
    train_data = list(zip(train_df["text"], train_df["category"]))
    test_data = list(zip(test_df["text"], test_df["category"]))
    classifier = NaiveBayesClassifier(train_data)
    acc = classifier.accuracy(test_data)
    print("Accuracy:", acc)


Fold 1
Train size: 1112
Test size: 1113
Accuracy: 0.9604672057502246

Fold 2
Train size: 1113
Test size: 1112
Accuracy: 0.9460431654676259


Task4


In [ ]:
import nltk
nltk.download('punkt')
from textblob.classifiers import NaiveBayesClassifier


def naiveBayesClassifier(trainingSet,testSet):
    classifier = NaiveBayesClassifier(trainingSet)
    correct = 0
    total = len(testSet)
    for doc, clas in testSet:
        predicted = classifier.classify(doc)
        print(f"document: ('{doc}', '{clas}')\t predicted class: {predicted}\n")
        if predicted == clas:
            correct += 1

    accuracy = correct / total
    print(f"Accuracy: {accuracy:.2f}")



trainingSet = [('London is the Capital of GB','GB'),
               ('Oxford is a city in GB','GB'),
               ('Dublin is the capital of Ireland','IE'),
               ('Limerick is a city in Ireland','IE')]
testSet = [('University of Limerick','IE'),
           ('University College Dublin','IE'),
           ('Imperial College London','GB'),
           ('University of Oxford','GB'),
           ('Ireland & GB','IE')]


naiveBayesClassifier(trainingSet,testSet)


document: ('University of Limerick', 'IE')	 predicted class: IE

document: ('University College Dublin', 'IE')	 predicted class: IE

document: ('Imperial College London', 'GB')	 predicted class: GB

document: ('University of Oxford', 'GB')	 predicted class: GB

document: ('Ireland & GB', 'IE')	 predicted class: IE

Accuracy: 1.00


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


 C. explain why the class of the last test document is predicted to be “GB” or “Ireland”

 Since the last document contain "Ireland" and "GB", they are present in both the classes,( Ireland in IE and GB in GB) hence the last document gets predicted as either "GB" or "IE". And the last document doesnt contain any another word to correctly predict it as either of those documents("&" doesnt belong to the dictionary).